File: app/csharp/tests/Ch04/HigherOrderFunctionsTests.cs
using Ch04;
using LanguageExt;
using static LanguageExt.Prelude;
using Xunit;

namespace GrokkingFP.CSharp.Tests.Ch04;

/// <summary>
/// 第4章: 高階関数のテスト
/// </summary>
public class HigherOrderFunctionsTests
{
    // ============================================
    // 基本的な高階関数のテスト
    // ============================================

    [Fact]
    public void WordScore_ExcludesLetterA()
    {
        Assert.Equal(3, HigherOrderFunctions.WordScore("Scala")); // "Scl" → 3
        Assert.Equal(8, HigherOrderFunctions.WordScore("function")); // no 'a' → 8
        Assert.Equal(0, HigherOrderFunctions.WordScore("aaa"));
    }

    [Fact]
    public void Bonus_Returns5WhenContainsC()
    {
        Assert.Equal(5, HigherOrderFunctions.Bonus("Scala"));
        Assert.Equal(0, HigherOrderFunctions.Bonus("Java"));
    }

    [Fact]
    public void Penalty_Returns7WhenContainsS()
    {
        Assert.Equal(7, HigherOrderFunctions.Penalty("Rust")); // lowercase 's'
        Assert.Equal(0, HigherOrderFunctions.Penalty("Java"));
    }

    [Fact]
    public void IsOdd_ReturnsTrueForOddNumbers()
    {
        Assert.True(HigherOrderFunctions.IsOdd(1));
        Assert.True(HigherOrderFunctions.IsOdd(3));
        Assert.False(HigherOrderFunctions.IsOdd(2));
    }

    [Fact]
    public void IsEven_ReturnsTrueForEvenNumbers()
    {
        Assert.True(HigherOrderFunctions.IsEven(0));
        Assert.True(HigherOrderFunctions.IsEven(2));
        Assert.False(HigherOrderFunctions.IsEven(1));
    }

    // ============================================
    // SortBy のテスト
    // ============================================

    [Fact]
    public void SortByScore_SortsAscending()
    {
        var words = Seq("ada", "haskell", "scala", "java");
        var sorted = HigherOrderFunctions.SortByScore(words);
        Assert.Equal(Seq("ada", "java", "scala", "haskell"), sorted);
    }

    [Fact]
    public void SortByScoreDescending_SortsDescending()
    {
        var words = Seq("ada", "haskell", "scala", "java");
        var sorted = HigherOrderFunctions.SortByScoreDescending(words);
        Assert.Equal(Seq("haskell", "scala", "java", "ada"), sorted);
    }

    // ============================================
    // Map のテスト
    // ============================================

    [Fact]
    public void Lengths_ReturnsLengthOfEachString()
    {
        var strings = Seq("hello", "world", "!");
        Assert.Equal(Seq(5, 5, 1), HigherOrderFunctions.Lengths(strings));
    }

    [Fact]
    public void Doubles_DoublesEachNumber()
    {
        var numbers = Seq(1, 2, 3, 4);
        Assert.Equal(Seq(2, 4, 6, 8), HigherOrderFunctions.Doubles(numbers));
    }

    // ============================================
    // Filter のテスト
    // ============================================

    [Fact]
    public void FilterOdds_ReturnsOnlyOddNumbers()
    {
        var numbers = Seq(1, 2, 3, 4, 5, 6);
        Assert.Equal(Seq(1, 3, 5), HigherOrderFunctions.FilterOdds(numbers));
    }

    [Fact]
    public void FilterLargerThan_ReturnsNumbersLargerThanN()
    {
        var numbers = Seq(1, 2, 3, 4, 5);
        Assert.Equal(Seq(4, 5), HigherOrderFunctions.FilterLargerThan(numbers, 3));
    }

    // ============================================
    // Fold のテスト
    // ============================================

    [Fact]
    public void Sum_ReturnsSumOfNumbers()
    {
        Assert.Equal(15, HigherOrderFunctions.Sum(Seq(1, 2, 3, 4, 5)));
        Assert.Equal(0, HigherOrderFunctions.Sum(Seq<int>()));
    }

    [Fact]
    public void Maximum_ReturnsMaximumNumber()
    {
        Assert.Equal(5, HigherOrderFunctions.Maximum(Seq(1, 5, 3, 2, 4)));
        Assert.Equal(-1, HigherOrderFunctions.Maximum(Seq(-5, -3, -1, -4)));
    }

    [Fact]
    public void CountWhere_CountsMatchingElements()
    {
        var numbers = Seq(1, 2, 3, 4, 5, 6);
        Assert.Equal(3, HigherOrderFunctions.CountWhere(numbers, HigherOrderFunctions.IsEven));
    }

    // ============================================
    // 関数を返す関数のテスト
    // ============================================

    [Fact]
    public void LargerThan_ReturnsPredicateFunction()
    {
        var largerThan5 = HigherOrderFunctions.LargerThan(5);
        Assert.True(largerThan5(6));
        Assert.False(largerThan5(5));
        Assert.False(largerThan5(4));
    }

    [Fact]
    public void DivisibleBy_ReturnsPredicateFunction()
    {
        var divisibleBy3 = HigherOrderFunctions.DivisibleBy(3);
        Assert.True(divisibleBy3(9));
        Assert.True(divisibleBy3(0));
        Assert.False(divisibleBy3(5));
    }

    [Fact]
    public void ContainsChar_ReturnsPredicateFunction()
    {
        var containsA = HigherOrderFunctions.ContainsChar('a');
        Assert.True(containsA("Scala"));
        Assert.False(containsA("PHP"));
    }

    // ============================================
    // ワードランキングのテスト
    // ============================================

    [Fact]
    public void RankByScore_RanksByBasicScore()
    {
        var words = Seq("ada", "haskell", "scala", "java");
        var ranked = HigherOrderFunctions.RankByScore(words);
        Assert.Equal("haskell", ranked.Head);
    }

    [Fact]
    public void RankByScoreWithBonus_IncludesBonus()
    {
        var words = Seq("ada", "haskell", "scala", "java");
        var ranked = HigherOrderFunctions.RankByScoreWithBonus(words);
        // scala: 3 + 5 = 8, haskell: 7 + 0 = 7
        Assert.Equal("scala", ranked.Head);
    }

    [Fact]
    public void RankByScoreWithBonusAndPenalty_IncludesBonusAndPenalty()
    {
        var words = Seq("ada", "haskell", "scala", "java");
        var ranked = HigherOrderFunctions.RankByScoreWithBonusAndPenalty(words);
        // haskell: 7 + 0 - 7 = 0
        // scala: 3 + 5 - 7 = 1
        // java: 2 + 0 - 0 = 2
        // ada: 0 + 0 - 0 = 0
        Assert.Equal("java", ranked.Head);
    }
}

/// <summary>
/// プログラミング言語のテスト
/// </summary>
public class ProgrammingLanguagesTests
{
    private readonly Seq<ProgrammingLanguages.ProgrammingLanguage> _languages = Seq(
        new ProgrammingLanguages.ProgrammingLanguage("Scala", 2004),
        new ProgrammingLanguages.ProgrammingLanguage("Haskell", 1990),
        new ProgrammingLanguages.ProgrammingLanguage("Java", 1995),
        new ProgrammingLanguages.ProgrammingLanguage("Rust", 2010)
    );

    [Fact]
    public void FilterByYear_ReturnsLanguagesAfterYear()
    {
        var filtered = ProgrammingLanguages.FilterByYear(_languages, 2000);
        Assert.Equal(2, filtered.Count);
        Assert.Contains(filtered, l => l.Name == "Scala");
        Assert.Contains(filtered, l => l.Name == "Rust");
    }

    [Fact]
    public void GetNames_ReturnsAllNames()
    {
        var names = ProgrammingLanguages.GetNames(_languages);
        Assert.Equal(Seq("Scala", "Haskell", "Java", "Rust"), names);
    }

    [Fact]
    public void SortByYear_SortsAscending()
    {
        var sorted = ProgrammingLanguages.SortByYear(_languages);
        var sortedList = sorted.ToList();
        Assert.Equal("Haskell", sortedList.First().Name);
        Assert.Equal("Rust", sortedList.Last().Name);
    }
}

File: app/fsharp/tests/Ch04/HigherOrderFunctionsTests.fs
module Ch04.HigherOrderFunctionsTests

open Xunit
open Ch04.HigherOrderFunctions
open Ch04.ProgrammingLanguages

// ============================================
// 基本的な関数のテスト
// ============================================

[<Fact>]
let ``wordScore は 'a' を除外した文字数を返す`` () =
    Assert.Equal(2, wordScore "java")   // j, v
    Assert.Equal(4, wordScore "rust")   // r, u, s, t
    Assert.Equal(3, wordScore "scala")  // s, c, l
    Assert.Equal(6, wordScore "haskell") // h, s, k, e, l, l

[<Fact>]
let ``bonus は 'c' を含む場合 5 を返す`` () =
    Assert.Equal(5, bonus "scala")
    Assert.Equal(5, bonus "clojure")
    Assert.Equal(0, bonus "java")
    Assert.Equal(0, bonus "rust")

[<Fact>]
let ``penalty は 's' を含む場合 7 を返す`` () =
    Assert.Equal(7, penalty "scala")
    Assert.Equal(7, penalty "rust")
    Assert.Equal(7, penalty "haskell")
    Assert.Equal(0, penalty "java")

// ============================================
// sortBy のテスト
// ============================================

[<Fact>]
let ``sortByScore はスコアの昇順でソートする`` () =
    let words = ["rust"; "java"]
    let sorted = sortByScore words

    // java: 2, rust: 4
    Assert.Equal<string list>(["java"; "rust"], sorted)

[<Fact>]
let ``sortByScoreDescending はスコアの降順でソートする`` () =
    let words = ["rust"; "java"; "haskell"]
    let sorted = sortByScoreDescending words

    // haskell: 6, rust: 4, java: 2
    Assert.Equal<string list>(["haskell"; "rust"; "java"], sorted)

// ============================================
// map のテスト
// ============================================

[<Fact>]
let ``lengths は各文字列の長さを返す`` () =
    Assert.Equal<int list>([5; 4; 3], lengths ["scala"; "rust"; "ada"])

[<Fact>]
let ``doubles は各数値を2倍にする`` () =
    Assert.Equal<int list>([10; 2; 4; 8; 0], doubles [5; 1; 2; 4; 0])

// ============================================
// filter のテスト
// ============================================

[<Fact>]
let ``filterOdds は奇数のみを返す`` () =
    Assert.Equal<int list>([5; 1], filterOdds [5; 1; 2; 4; 0])

[<Fact>]
let ``filterLargerThan は指定した値より大きい要素を返す`` () =
    Assert.Equal<int list>([5], filterLargerThan 4 [5; 1; 2; 4; 0])
    Assert.Equal<int list>([5; 2; 4], filterLargerThan 1 [5; 1; 2; 4; 0])

// ============================================
// fold のテスト
// ============================================

[<Fact>]
let ``sum はリストの合計を返す`` () =
    Assert.Equal(112, sum [5; 1; 2; 4; 100])
    Assert.Equal(0, sum [])

[<Fact>]
let ``maximum はリストの最大値を返す`` () =
    Assert.Equal(15, maximum [5; 1; 2; 4; 15])
    Assert.Equal(100, maximum [5; 1; 100; 4; 15])

[<Fact>]
let ``countWhere は条件を満たす要素の数を返す`` () =
    Assert.Equal(2, countWhere (fun i -> i > 3) [1; 2; 3; 4; 5])
    Assert.Equal(2, countWhere (fun (s: string) -> s.Length > 1) ["a"; "bb"; "ccc"])

// ============================================
// 関数を返す関数のテスト
// ============================================

[<Fact>]
let ``largerThan はnより大きいか判定する関数を返す`` () =
    let largerThan4 = largerThan 4

    Assert.True(largerThan4 5)
    Assert.False(largerThan4 4)
    Assert.False(largerThan4 3)

[<Fact>]
let ``divisibleBy はnで割り切れるか判定する関数を返す`` () =
    let divisibleBy3 = divisibleBy 3

    Assert.True(divisibleBy3 9)
    Assert.True(divisibleBy3 6)
    Assert.False(divisibleBy3 7)

[<Fact>]
let ``containsChar は指定した文字を含むか判定する関数を返す`` () =
    let containsA = containsChar 'a'

    Assert.True(containsA "scala")
    Assert.True(containsA "java")
    Assert.False(containsA "rust")

// ============================================
// ワードランキングのテスト
// ============================================

[<Fact>]
let ``rankedWords は指定したスコア関数でランキングする`` () =
    let words = ["ada"; "haskell"; "scala"; "java"; "rust"]

    // 基本スコアでランキング (haskell:6, rust:4, scala:3, java:2, ada:1)
    let ranking1 = rankByScore words
    Assert.Equal<string list>(["haskell"; "rust"; "scala"; "java"; "ada"], ranking1)

[<Fact>]
let ``rankByScoreWithBonus はボーナス付きでランキングする`` () =
    let words = ["ada"; "haskell"; "scala"; "java"; "rust"]

    // ボーナス付きスコアでランキング
    // scala: 3+5=8, haskell: 6+0=6, rust: 4+0=4, java: 2+0=2, ada: 1+0=1
    let ranking2 = rankByScoreWithBonus words
    Assert.Equal<string list>(["scala"; "haskell"; "rust"; "java"; "ada"], ranking2)

[<Fact>]
let ``rankByScoreWithBonusAndPenalty はボーナスとペナルティ付きでランキングする`` () =
    let words = ["ada"; "haskell"; "scala"; "java"; "rust"]

    // ボーナスとペナルティ付きスコアでランキング
    // java: 2+0-0=2, scala: 3+5-7=1, ada: 1+0-0=1, haskell: 6+0-7=-1, rust: 4+0-7=-3
    // scala と ada は同スコアなので順序は安定ソートに依存
    let ranking3 = rankByScoreWithBonusAndPenalty words
    // java が最高スコアで最初、haskell と rust が最後
    Assert.Equal("java", ranking3.[0])
    Assert.Equal("haskell", ranking3.[3])
    Assert.Equal("rust", ranking3.[4])

// ============================================
// ProgrammingLanguages のテスト
// ============================================

[<Fact>]
let ``getNames は言語名のリストを返す`` () =
    let languages = [
        { Name = "Java"; Year = 1995 }
        { Name = "Scala"; Year = 2004 }
    ]

    Assert.Equal<string list>(["Java"; "Scala"], getNames languages)

[<Fact>]
let ``filterByYear は指定した年より後の言語をフィルタする`` () =
    let languages = [
        { Name = "Java"; Year = 1995 }
        { Name = "Scala"; Year = 2004 }
        { Name = "Kotlin"; Year = 2011 }
    ]

    let young = filterByYear 2000 languages

    Assert.Equal(2, List.length young)
    Assert.Equal<string list>(["Scala"; "Kotlin"], getNames young)

[<Fact>]
let ``sortByYear は年でソートする`` () =
    let languages = [
        { Name = "Scala"; Year = 2004 }
        { Name = "Java"; Year = 1995 }
        { Name = "Kotlin"; Year = 2011 }
    ]

    let sorted = sortByYear languages

    Assert.Equal<string list>(["Java"; "Scala"; "Kotlin"], getNames sorted)
